File: app/src/main/java/com/orgzly/android/repos/VersionedRook.java
package com.orgzly.android.repos;

import android.net.Uri;

/**
 * Versioned {@link Rook}.
 *
 * A specific version of {@link Rook} with revision and modification time.
 */
public class VersionedRook extends Rook {
    private String revision;
    private long mtime;

    public VersionedRook(Rook rook, String revision, long mtime) {
        this(rook.getRepoUri(), rook.getUri(), revision, mtime);
    }

    public VersionedRook(Uri repoUri, Uri uri, String revision, long mtime) {
        super(repoUri, uri);

        this.revision = revision;
        this.mtime = mtime;
    }

    public String getRevision() {
        return revision;
    }

    public long getMtime() {
        return mtime;
    }

    public String toString() {
        return uri.buildUpon()
                .appendQueryParameter("revision", revision)
                .appendQueryParameter("mtime", String.valueOf(mtime))
                .build().toString();
    }
}

File: app/src/main/java/com/orgzly/android/repos/VersionedRook.kt
package com.orgzly.android.repos

import android.net.Uri

class VersionedRook(
        repoId: Long,
        repoType: RepoType,
        repoUri: Uri,
        uri: Uri,
        val revision: String,
        val mtime: Long
) : Rook(repoId, repoType, repoUri, uri) {

    constructor(rook: Rook, revision: String, mtime: Long) :
            this(rook.repoId, rook.repoType, rook.getRepoUri(), rook.getUri(), revision, mtime)

    override fun toString(): String {
        return uri.buildUpon()
                .appendQueryParameter("revision", revision)
                .appendQueryParameter("mtime", mtime.toString())
                .build().toString()
    }
}
